### PR TITLE
54 trading

### PR DESCRIPTION
**Pull request by Ghost-Element**

### CODE REVIEW

File: app/src/test/java/com/group2/catan_android/data/repository/TradeRepositoryTest.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+    @Test
+    void testCorrectTradeOfferDtoGetsSend() {
+        TestObserver<TradeOfferDto> testObserver = tradeRepository.getTradeObservable().test();
+
+        TradeOfferDto dto = new TradeOfferDto(new int[]{1, 2, 3, 4, 5}, new int[]{5, 4, 3, 2, 1}, new IngamePlayerDto("", new int[]{0, 0, 0, 0, 0}, 3, 7, 3, new ArrayList<>()));
+        liveIn.onNext(dto);
+        dto = testObserver.values().get(testObserver.values().size() - 1);
+
+        Assertions.assertArrayEquals(new int[]{5, 4, 3, 2, 1}, dto.getGetResources());
```
Comment (jastornig) [inline app/src/test/java/com/group2/catan_android/data/repository/TradeRepositoryTest.java:49]: don't forget to rename the dto fields. getRessources, giveRessources is not nice.
offered, wanted ore something like this is better. See the previous pr for suggestions.

File: app/src/main/java/com/group2/catan_android/fragments/TradingResourceSelectionFragment.java [hunk 2]
```diff
@@ -64,51 +70,19 @@ public void setPlusAndMinusListener(View view){
         for(int i=0;i<plus.length;i++){
             final int j = i;
             plus[j].setOnClickListener(v -> {
-                int num = getNumberofTextView(count[j]);
-                if (num == -1) {
-                    return;
-                }//Error
-                count[j].setText(Integer.toString(++num));
+                count[j].setText(Integer.toString(tradeResourcesViewModel.togglePlus(j)));
```
Comment (jastornig) [inline app/src/main/java/com/group2/catan_android/fragments/TradingResourceSelectionFragment.java:73]: Do what the compiler says. Also in general, if a warning is shown, make it go away

File: app/src/main/java/com/group2/catan_android/data/live/game/AcceptTradeOfferMoveDto.java [hunk 3]
```diff
@@ -1,12 +1,12 @@
 package com.group2.catan_android.data.live.game;
-public class AcceptMoveDto extends GameMoveDto {
-    public AcceptMoveDto() {
-        this.setEventType(GameMoveType.ACCEPTMOVE);
+public class AcceptTradeOfferMoveDto extends GameMoveDto {
+    public AcceptTradeOfferMoveDto() {
+        this.setEventType(GameMoveType.ACCEPTRADETMOVE);
```
Comment (jastornig) [inline app/src/main/java/com/group2/catan_android/data/live/game/AcceptTradeOfferMoveDto.java:4]: Typo

**Review by jastornig — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/group2/catan_android/data/live/game/GameMoveType.java [hunk 4]
```diff
@@ -10,6 +10,6 @@ public class GameMoveType {
     public static final String BUYPROGRESSCARD = "BUY_PROGRESS_CARD";
     public static final String MOVEROBBERMOVE = "MOVE_ROBBER_MOVE";
     public static final String ACCUSECHEATINGMOVE="ACCUSE_CHEATING_MOVE";
-    public static final String TRADEMOVE = "TRADE_MOVE";
-    public static final String ACCEPTMOVE = "ACCEPT_MOVE";
+    public static final String MAKETRADEMOVE = "MAKE_TRADE_MOVE";
+    public static final String ACCEPTRADETMOVE = "ACCEPT_TRADE_MOVE";
```
[the remark this answers is not in the document]
Comment (Ghost-Element) [reply app/src/main/java/com/group2/catan_android/data/live/game/GameMoveType.java:14]: changed to ACCEPTTRADEMOVE

File: app/src/main/java/com/group2/catan_android/fragments/TradingResourceSelectionFragment.java [hunk 5]
```diff
@@ -64,51 +70,19 @@ public void setPlusAndMinusListener(View view){
...
-                int num = getNumberofTextView(count[j]);
-                if (num == -1) {
-                    return;
-                }//Error
-                num--;
-                if (num < 0) {
-                    return;
+                int result= tradeResourcesViewModel.toggleMinus(j);
+                if(result<=-1){
+                    MessageBanner.makeBanner(requireActivity(), MessageType.ERROR, "Clicked Resource can not me less than zero!").show();
```
[the remark this answers is not in the document]
Comment (Ghost-Element) [reply app/src/main/java/com/group2/catan_android/fragments/TradingResourceSelectionFragment.java:78]: yes the user should know that

File: app/src/main/java/com/group2/catan_android/viewmodel/TradeResourcesViewModel.java [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+        return ++resources[index];
+    }
+    public int toggleMinus(int index){
+        if(resources[index]==0){
+            return -1;
+        }else {
+            return --resources[index];
+        }
+    }
+    public int[] getResources(){
```
[the remark this answers is not in the document]
Comment (Ghost-Element) [reply app/src/main/java/com/group2/catan_android/viewmodel/TradeResourcesViewModel.java:26]: inserted Indexes of Resources of Logic-part as constants and used them

File: app/src/test/java/com/group2/catan_android/data/repository/TradeRepositoryTest.java (hunk 1 above)
Comment (Ghost-Element) [reply app/src/test/java/com/group2/catan_android/data/repository/TradeRepositoryTest.java:49]: refactored accoringly

File: app/src/main/java/com/group2/catan_android/data/live/game/AcceptTradeOfferMoveDto.java (hunk 3 above)
Comment (Ghost-Element) [reply app/src/main/java/com/group2/catan_android/data/live/game/AcceptTradeOfferMoveDto.java]: fixed

File: app/src/main/java/com/group2/catan_android/fragments/TradingResourceSelectionFragment.java (hunk 2 above)
Comment (Ghost-Element) [reply app/src/main/java/com/group2/catan_android/fragments/TradingResourceSelectionFragment.java]: changed to String.format()

File: app/src/main/java/com/group2/catan_android/viewmodel/TradeResourcesViewModel.java [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+
+public class TradeResourcesViewModel extends ViewModel {
+
+    private int[] resources;
+
+    public TradeResourcesViewModel(){
+        resources = new int[]{0,0,0,0,0};
+    }
+
+    public int togglePlus(int index){
```
[the remark this answers is not in the document]
Comment (Ghost-Element) [reply app/src/main/java/com/group2/catan_android/viewmodel/TradeResourcesViewModel.java]: changed as LiveData

**Review by jastornig — APPROVED** [review]

**Pull request merged**